core/src/test/scala/unit/kafka/server/UpdateFeaturesTest.scala

```diff
@@ -0,0 +1,550 @@
+        Utils.mkEntry("feature_2", new FinalizedVersionRange(2, 3))))
+    val update1 = new FeatureUpdate(targetFinalizedFeatures.get("feature_1").max(), false)
+    val update2 = new FeatureUpdate(targetFinalizedFeatures.get("feature_2").max(), false)
+
+    val expected = new FeatureMetadata(
+      targetFinalizedFeatures,
+      versionBefore + 1,
+      Features.supportedFeatures(
+        Utils.mkMap(
+          Utils.mkEntry("feature_1", new SupportedVersionRange(1, 3)),
+          Utils.mkEntry("feature_2", new SupportedVersionRange(2, 5)))))
+
+    val adminClient = createAdminClient()
+    adminClient.updateFeatures(
+      Utils.mkMap(Utils.mkEntry("feature_1", update1), Utils.mkEntry("feature_2", update2)),
+      new UpdateFeaturesOptions()
+    ).all().get()
+
+    checkFeatures(
+      adminClient,
+      new FeatureZNode(FeatureZNodeStatus.Enabled, targetFinalizedFeatures),
+      expected)
+  }
+
+  @Test
+  def testSuccessfulFeatureUpgradeAndDowngrade(): Unit = {
+    TestUtils.waitUntilControllerElected(zkClient)
+
+    updateSupportedFeaturesInAllBrokers(
+      Features.supportedFeatures(
+        Utils.mkMap(
+          Utils.mkEntry("feature_1", new SupportedVersionRange(1, 3)),
+          Utils.mkEntry("feature_2", new SupportedVersionRange(2, 5)))))
+    updateDefaultMinVersionLevelsInAllBrokers(Map[String, Short]("feature_1" -> 1, "feature_2" -> 2))
+    val versionBefore = updateFeatureZNode(
+      Features.finalizedFeatures(
+        Utils.mkMap(
+          Utils.mkEntry("feature_1", new FinalizedVersionRange(1, 2)),
+          Utils.mkEntry("feature_2", new FinalizedVersionRange(2, 4)))))
+
+    val targetFinalizedFeatures = Features.finalizedFeatures(
+      Utils.mkMap(
+        Utils.mkEntry("feature_1", new FinalizedVersionRange(1, 3)),
+        Utils.mkEntry("feature_2", new FinalizedVersionRange(2, 3))))
+    val update1 = new FeatureUpdate(targetFinalizedFeatures.get("feature_1").max(), false)
+    val update2 = new FeatureUpdate(targetFinalizedFeatures.get("feature_2").max(), true)
+
+    val expected = new FeatureMetadata(
+      targetFinalizedFeatures,
+      versionBefore + 1,
+      Features.supportedFeatures(
+        Utils.mkMap(
+          Utils.mkEntry("feature_1", new SupportedVersionRange(1, 3)),
+          Utils.mkEntry("feature_2", new SupportedVersionRange(2, 5)))))
+
+    val adminClient = createAdminClient()
+    adminClient.updateFeatures(
+      Utils.mkMap(Utils.mkEntry("feature_1", update1), Utils.mkEntry("feature_2", update2)),
+      new UpdateFeaturesOptions()
+    ).all().get()
+
+    checkFeatures(
+      adminClient,
+      new FeatureZNode(FeatureZNodeStatus.Enabled, targetFinalizedFeatures),
+      expected)
+  }
+
+  @Test
+  def testPartialSuccessDuringValidFeatureUpgradeAndInvalidDowngrade(): Unit = {
+    TestUtils.waitUntilControllerElected(zkClient)
+
+    val initialSupportedFeatures = Features.supportedFeatures(
```

Comment:
nit: this could be extracted as a common struct.

Reply:
Just 2 occurrences (one in this test and other in the next test). I'd leave it the way it is as the test is readable with values inlined in the test body.